### ISSUE TITLE
ASE_ASSERTION_WITH_SIDE_EFFECT_METHOD produce false positives

### ISSUE THREAD

**Issue by dougxc** [spotbugs/spotbugs#3483]
I've been trying SpotBugs on the Graal compiler code using a snapshot build, running on a JDK 25 early access build. The latter is required since Graal references some recent JDK internals (e.g. [CompilerThreadCanCallJavaScope](https://github.com/openjdk/jdk/blob/master/src/jdk.internal.vm.ci/share/classes/jdk/vm/ci/hotspot/CompilerThreadCanCallJavaScope.java)) and it seems as though SpotBugs relies on the underlying JDK to resolve and analyze such accesses.

However, it seems as though SpotBugs is producing some false positive in my setup. Since I don't know where the problem actually is, I can only provide steps for reproducing the issue.

```
# Set JAVA_HOME to JDK 17
./gradlew clean
./gradlew build
./gradlew distZip
unzip ./spotbugs/build/distributions/spotbugs-4.9.4-SNAPSHOT.zip
export SPOTBUGS_HOME=$PWD/spotbugs-4.9.4-SNAPSHOT
export MX_PYTHON=python3 # Ensure this is python 3.8 or later
git clone https://github.com/graalvm/mx.git
export PATH=$PATH:$PWD/mx
git clone https://github.com/oracle/graal.git
mx -p graal/compiler fetch-jdk labsjdk-ce-latest --to . --auto-alias --strip-contents-home
export JAVA_HOME=$PWD/labsjdk-ce-latest
mx -p graal/compiler build
mx -p graal/compiler spotbugs
```

This should produce output that includes warnings such as:
```
L S ASE: The method called in assertion at jdk.graal.compiler.libgraal.truffle.LibGraalTruffleEntryPoints.getCompilationId(JNI$JNIEnv, JNI$JClass, long, long) may produce a side effect. If assertions are disabled, the method invocation won't get executed and the result of the method might change.  At LibGraalTruffleEntryPoints.java:[line 330]
```

The above ASE warning cannot be right as there is no `assert` anywhere in [LibGraalTruffleEntryPoints.java](https://github.com/oracle/graal/blob/master/compiler/src/jdk.graal.compiler.libgraal/src/jdk/graal/compiler/libgraal/truffle/LibGraalTruffleEntryPoints.java).

**Comment by gtoison** [comment]
Can you check (or share) the `.class` file?
I guess that the Graal compiler could potentially add some asserts for methods annotated with `@CEntryPoint`

**Comment by dougxc** [comment]
> Can you check (or share) the `.class` file? I guess that the Graal compiler could potentially add some asserts for methods annotated with `@CEntryPoint`

Here's the `javap -verbose` disassembly of `LibGraalTruffleEntryPoints.class`:
[LibGraalTruffleEntryPoints.class.txt](https://github.com/user-attachments/files/20677305/LibGraalTruffleEntryPoints.class.txt)

Note that the Graal compiler is a JIT compiler written in java so does not produce class files - it reads them and compiles them to native code.

**Comment by gtoison** [comment]
Thanks, I definitively need to educate myself about Graal 😄 
I looked a bit into how SpotBugs detects `ASE_ASSERTION_WITH_SIDE_EFFECT_METHOD` and it seems to be looking for "setter like" method calls on a mutable class, between the `$assertionsDisabled` load and the `new AssertionError();`

If, for some reason, it does not see `new AssertionError();` the `inAssert` flag in [AbstractAssertDetector](https://github.com/spotbugs/spotbugs/blob/master/spotbugs/src/main/java/edu/umd/cs/findbugs/detect/AbstractAssertDetector.java) could potentially fail to reset to false and the detector might still act as if it was inside an assert.

It's going to be hard to diagnose this without reproducing, but in the meantime you could try add the following method in `AbstractAssertDetector`:

```
    @Override
    public void visitAfter(Code obj) {
        inAssert = false;
    }
```

**Comment by gtoison** [comment]
I've pushed a commit here: https://github.com/gtoison/spotbugs/tree/issue-3483
I don't think the change will break something but at this point it's only a wild guess

**Comment by dougxc** [comment]
Thanks for the patch @gtoison but unfortunately it does not appear to fix the problem.

**Comment by gtoison** [comment]
Thanks for trying this out!
I attempted to build graal myself but seem to run into issue with the native part, I get this error in the build report:
```
Building org.graalvm.nativebridge.launcher_windows-amd64-default with Ninja... [dependencies were added, removed or re-ordered]

[1/2] CXX src\launcher.obj

FAILED: src/launcher.obj

cl -nologo -showIncludes -IC:\dev\eclipse-workspace-graal\graal\labsjdk-ce-latest-25+26-jvmci-b01\include -IC:\dev\eclipse-workspace-graal\graal\labsjdk-ce-latest-25+26-jvmci-b01\include\win32  /std:c++17 -c C:\dev\eclipse-workspace-graal\graal\sdk\src\org.graalvm.nativebridge.launcher\src\launcher.cc -Fosrc\launcher.obj

CreateProcess failed: The system cannot find the file specified.

ninja: build stopped: subcommand failed.

Building org.graalvm.nativebridge.launcher_windows-amd64-default with Ninja: Failed due to error: 1
```

And:
```
Testing org.graalvm.toolchain.test... [dependencies were added, removed or re-ordered]

rerunning with -v

clang version 20.1.4 (GraalVM.org llvmorg-20.1.4-1-ga7183f5a17-bg217527b869 a7183f5a171aab1c57194db203ee0dea078a2570)

Target: x86_64-pc-windows-msvc

Thread model: posix

InstalledDir: C:\dev\eclipse-workspace-graal\graal\sdk\mxbuild\windows-amd64\LLVM_TOOLCHAIN\bin

 "C:\\dev\\eclipse-workspace-graal\\graal\\sdk\\mxbuild\\windows-amd64\\LLVM_TOOLCHAIN\\bin\\clang.exe" -cc1 -triple x86_64-pc-windows-msvc19.33.0 -emit-obj -mincremental-linker-compatible -dumpdir "C:\\dev\\eclipse-workspace-graal\\graal\\sdk\\mxbuild\\windows-amd64\\org.graalvm.toolchain.test\\hello.exe-" -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name hello.c -mrelocation-model pic -pic-level 2 -mframe-pointer=none -relaxed-aliasing -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -fms-volatile -funwind-tables=2 -target-cpu x86-64 -tune-cpu generic "-fdebug-compilation-dir=C:\\dev\\eclipse-workspace-graal" -v "-fcoverage-compilation-dir=C:\\dev\\eclipse-workspace-graal" -resource-dir "C:\\dev\\eclipse-workspace-graal\\graal\\sdk\\mxbuild\\windows-amd64\\LLVM_TOOLCHAIN\\lib\\clang\\20" -internal-isystem "C:\\dev\\eclipse-workspace-graal\\graal\\sdk\\mxbuild\\windows-amd64\\LLVM_TOOLCHAIN\\lib\\clang\\20\\include" -internal-isystem "C:/Program Files/Microsoft Visual Studio 10.0/VC/include" -internal-isystem "C:/Program Files/Microsoft Visual Studio 9.0/VC/include" -internal-isystem "C:/Program Files/Microsoft Visual Studio 9.0/VC/PlatformSDK/Include" -internal-isystem "C:/Program Files/Microsoft Visual Studio 8/VC/include" -internal-isystem "C:/Program Files/Microsoft Visual Studio 8/VC/PlatformSDK/Include" -ferror-limit 19 -fno-use-cxa-atexit -fms-extensions -fms-compatibility -fms-compatibility-version=19.33 -fskip-odr-check-in-gmf -fdelayed-template-parsing -faddrsig -o "C:\\Users\\gtoison\\AppData\\Local\\Temp\\hello-97ce47.o" -x c "C:\\dev\\eclipse-workspace-graal\\graal\\sdk\\src\\org.graalvm.toolchain.test\\hello.c"

clang -cc1 version 20.1.4 based upon LLVM 20.1.4 default target x86_64-pc-windows-msvc

ignoring nonexistent directory "C:/Program Files/Microsoft Visual Studio 10.0/VC/include"

ignoring nonexistent directory "C:/Program Files/Microsoft Visual Studio 9.0/VC/include"

ignoring nonexistent directory "C:/Program Files/Microsoft Visual Studio 9.0/VC/PlatformSDK/Include"

ignoring nonexistent directory "C:/Program Files/Microsoft Visual Studio 8/VC/include"

ignoring nonexistent directory "C:/Program Files/Microsoft Visual Studio 8/VC/PlatformSDK/Include"

#include "..." search starts here:

#include <...> search starts here:

 C:\dev\eclipse-workspace-graal\graal\sdk\mxbuild\windows-amd64\LLVM_TOOLCHAIN\lib\clang\20\include

End of search list.

C:\dev\eclipse-workspace-graal\graal\sdk\src\org.graalvm.toolchain.test\hello.c:41:10: fatal error: 'stdio.h' file not found

   41 | #include <stdio.h>

      |          ^~~~~~~~~

1 error generated.

The LLVM toolchain does not work. Do you have development packages installed?

Testing org.graalvm.toolchain.test: Failed due to error: 1
```

Nonetheless this seems to have compiled some java file because I can see `.class` files so I attempted to run spotbugs but got an other error:
```
C:\dev\eclipse-workspace-graal>mx -p graal/compiler spotbugs
WARNING: symlinking not supported
Traceback (most recent call last):
  File "C:\dev\eclipse-workspace-graal\mx\mx.py", line 31, in <module>
    runpy.run_module("mx")
  File "C:\Users\gtoison\AppData\Local\Programs\Python\Python310\lib\runpy.py", line 227, in run_module
    return _run_code(code, {}, init_globals, run_name, mod_spec)
  File "C:\Users\gtoison\AppData\Local\Programs\Python\Python310\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\dev\eclipse-workspace-graal\mx\src\mx\__main__.py", line 49, in <module>
    _main_wrapper()
  File "C:\dev\eclipse-workspace-graal\mx\src\mx\_impl\mx.py", line 18602, in _main_wrapper
    main()
  File "C:\dev\eclipse-workspace-graal\mx\src\mx\_impl\mx.py", line 18586, in main
    retcode = c(command_args)
  File "C:\dev\eclipse-workspace-graal\mx\src\mx\_impl\mx_commands.py", line 145, in __call__
    return self.command_function(*args, **kwargs)
  File "C:\dev\eclipse-workspace-graal\mx\src\mx\_impl\mx_spotbugs.py", line 128, in spotbugs
    resultcode = max(resultcode, _spotbugs(args, spotbugsArgs, suite, versionProjects, spotbugsVersion))
  File "C:\dev\eclipse-workspace-graal\mx\src\mx\_impl\mx_spotbugs.py", line 211, in _spotbugs
    jdk = mx.get_jdk() if spotbugsHome else mx.get_tools_jdk(_range, purpose='SpotBugs')
  File "C:\dev\eclipse-workspace-graal\mx\src\mx\_impl\mx.py", line 13001, in get_tools_jdk
    _tools_jdks[cache_key] = get_jdk(versionCheck=versionCheck, purpose=purpose, cancel=cancel, versionDescription=versionDescription, tag=tag, abortCallback=_dummy_abort, **kwargs)
  File "C:\dev\eclipse-workspace-graal\mx\src\mx\_impl\mx.py", line 12943, in get_jdk
    abort_not_found()
  File "C:\dev\eclipse-workspace-graal\mx\src\mx\_impl\mx.py", line 12917, in abort_not_found
    abortCallback(msg)
  File "C:\dev\eclipse-workspace-graal\mx\src\mx\_impl\mx.py", line 12999, in _dummy_abort
    abort_not_found()
  File "C:\dev\eclipse-workspace-graal\mx\src\mx\_impl\mx.py", line 12977, in abort_not_found
    import select_jdk
  File "C:\dev\eclipse-workspace-graal\mx\select_jdk.py", line 1, in <module>
    src/mx/_impl/select_jdk.py
NameError: name 'src' is not defined
```

There's probably something obvious I'm missing but I'm not too used to work with C or Python.
Or maybe you can share the build folder (SpotBugs should only read the .class files) and I can try reproducing from there?

**Comment by dougxc** [comment]
Oh no, Windows... ;-) Getting Graal development set up on Windows is a bit painful and I'm not sure it worth trying to figure it out here.

I've tried to boil this down to a more limited reproducer that avoids the native build steps but have not had any luck yet. I'll post more here if/when I figure it out.

**Comment by gtoison** [comment]
Thanks, you can also try to throw an exception instead of resetting the `inAssert` in `visitAfter(Code obj)`
Again that's just a guess but apparently the expectation of the author was to always see `new AssertionError()`
So this could potentially help checking if the expectation was correct:
```
    @Override
    public void visitAfter(Code obj) {
    	if (inAssert) {
    		throw new RuntimeException("Shouldn't be here: " + getClassName() + ": " + getMethodName());
    	}
    }
```

**Comment by woess** [comment]
Some reproducers for `ASE_ASSERTION_WITH_SIDE_EFFECT` false positives:
```java
public class MethodBodyFalsePositive {
    public static boolean $assertionsDisabled;
    private static int field;

    public static void poison() {
        if ($assertionsDisabled) {
            return;
        }
    }

    public static void victim() {
        field = 42;
    }
}
```
```
java -jar spotbugs-4.9.8.jar -textui -low -maxRank 15 MethodBodyFalsePositive.class
L S ASE: The expression used in assertion at MethodBodyFalsePositive.victim() may produce a side effect. If assertions are disabled, the expression won't get executed and the result of the method might change.  At MethodBodyFalsePositive.java:[line 12]
```

```java
public class ConstructorFalsePositive {
    public static boolean $assertionsDisabled;
    private final Object first;
    private final Object second;

    public static void poison() {
        if ($assertionsDisabled) {
            return;
        }
    }

    public ConstructorFalsePositive(Object first, Object second) {
        this.first = first;
        this.second = second;
    }
}
```
```
java -jar spotbugs-4.9.8.jar -textui -low -maxRank 15 ConstructorFalsePositive.class
L S ASE: The expression used in assertion at new ConstructorFalsePositive(Object, Object) may produce a side effect. If assertions are disabled, the expression won't get executed and the result of the method might change.  At ConstructorFalsePositive.java:[line 13]
L S ASE: The expression used in assertion at new ConstructorFalsePositive(Object, Object) may produce a side effect. If assertions are disabled, the expression won't get executed and the result of the method might change.  At ConstructorFalsePositive.java:[line 14]
```

```java
public class Poison {
    public static boolean $assertionsDisabled;
    public static void poison() { if ($assertionsDisabled) return; }
}
public class Victim {
    static int field;
    public static void victim() { field = 1; }
}
```
```
java -jar spotbugs-4.9.8.jar -textui -low -maxRank 15 Poison.class Victim.class
L S ASE: The expression used in assertion at Victim.victim() may produce a side effect. If assertions are disabled, the expression won't get executed and the result of the method might change.  At Victim.java:[line 3]
```

**Comment by JuditKnoll** [comment]
From #4242 PR description: 
> @dougxc said the visitAfter patch did not fix the Graal build, and I could not reproduce that setup. This PR fixes the reproducers @woess posted, so I have deliberately not written "closes https://github.com/spotbugs/spotbugs/issues/3483"

GitHub recognized this and closed the issue, so I'm reopening it.
